src/middlewared/middlewared/main.py

```diff
@@ -898,18 +905,128 @@ def pipe(self):
         if job:
             return job
         else:
+
+            # Currently its only a boolean
+            if serviceobj._config.process_pool is True:
+                return await self.run_in_thread(self._call_worker, serviceobj, name, *args)
+
             if asyncio.iscoroutinefunction(methodobj):
                 return await methodobj(*args)
             else:
-                pool = None
+                tpool = None
                 if serviceobj._config.thread_pool:
-                    pool = serviceobj._config.thread_pool
+                    tpool = serviceobj._config.thread_pool
                 if hasattr(methodobj, '_thread_pool'):
-                    pool = methodobj._thread_pool
-                if pool:
-                    return await self.run_in_thread_pool(pool, methodobj, *args)
-                else:
-                    return await self.run_in_thread(methodobj, *args)
+                    tpool = methodobj._thread_pool
+                if tpool:
+                    return await self.run_in_executor(tpool, methodobj, *args)
+
+                return await self.run_in_thread(methodobj, *args)
+
+    def _call_worker(self, serviceobj, name, *args, job=None):
```

Comment:
`run_in_proc` method added but never used. Is there any problem with `ProcessPoolExecutor`? Wouldn't it be more interesting and rewarding to solve them instead of implementing the very same thing ourself? I feel this code is complicated and may contain obscure bugs. At least it should be in another module. 100+ lines of low-level code that has little to do with job handling logic in our middleware.

Reply:
ProcessPoolExecutor cannot run middleware plugins directly. It needs to be able to pickle all related classes and that is not possible because of Service attributes.

Am I missing something?

Reply:
@themylogin the `run_in_proc` I just added for convenience, there are some calls to libzfs in another module that could be a problem and I had plans to use it. Do you think it should be removed for now?

Reply:
@william-gr I'll try to use `ProcessPoolExecutor` to run middleware plugin later today, I think this can be done. Can you wait a few hours before merging this please?

Reply:
@themylogin I want to do it myself if you dont mind.

I know ProcessPoolExecutor can be used, just not directly. We just need to make components that can be pickled.

I can try to use ProcessPool but then we lose control of things like stdout/stderr of the process, possibility of extracting resource usage and the possibility of having the worker as a long running process in the future. I imagine this will become a thing someday because of python GIL and CPU intensive tasks (e.g. replication) and that can't happen on same process.

Any thoughts on this, do you think its a wrong assessment?

Reply:
Oh and partially because spawning a new process everytime is resource intensive (starting up python and loading all the modules every call can take time)

Reply:
> We just need to make components that can be pickled.

I thought of trying to avoid passing components to the `ProcessPoolExecutor`. Only pass class name and method name instead. Components are stateless so this should work.

Reply:
I thought of trying to avoid passing components to the ProcessPoolExecutor. Only pass class 
    name and method name instead. Components are stateless so this should work.```

Right, but you need to be able to make calls between methods (e.g. use self.middleware.call(_sync)?), so the middleware client needs to be instantiated, and well as manually loading the modules, some wrapping will be necessary. (Just like worker.py)

What about the other points I raised, you think they are not right/useful/necessary?

Reply:
> so the middleware client needs to be instantiated, and well as manually loading the modules, some wrapping will be necessary. (Just like worker.py)

I don't see any issue with this if we use process pool. Initialization will only happen once.

> we lose control of things like stdout/stderr of the process

Isn't it inherited?

> possibility of extracting resource usage

Is this done already? How?

> possibility of having the worker as a long running process in the future

`concurrent.futures.ProcessPoolExecutor` maintains the process pool of long-running processes, isn't it?

Reply:
```Isn't it inherited?```

Thats precisely the issue, how do you control or get that separately

```Is this done already? How?```

Not yet, but its fairly easy to grab before call ends.

```concurrent.futures.ProcessPoolExecutor maintains the process pool of long-running processes, isn't it?```

It is long-running process but I dont understand how that works, perhaps you can explain to me. As far as I understand process pool spawns a new python process but it runs its own python method to communicate with the parent process. How can we have control of that new process to load things globally and retain the state we want?

Reply:
> Thats precisely the issue, how do you control or get that separately

Why do we want to control it separately?

> How can we have control of that new process to load things globally and retain the state we want?

This process is created using `fork`. Won't it retain everything we need?

Reply:
`Why do we want to control it separately?`

To show/preserve stdout/stderr (e.g. iocage and other calls might print things to stdout/stderr)

`This process is created using fork. Won't it retain everything we need?`

So it will be a bloated process which has everything, include the running asyncio event loop?